### PR TITLE
Add a new fusion instruction and remove unnecessary ones 

### LINKED ISSUE #226: Constant propagation and constant folding
Based on our observation, RISC-V architecture need two instruction `auipc + addi` to create a constant. If we apply constant propagation and constant folding, we could replace them with a simple instruction `lui`. Moreover, these two optimizations can apply in other scenarioes. For example, any instruction uses the constant created by `auipc + addi` can be optimized.

If you want to move 0xffffffff to a register, all the assembly code you need is:
```
LUI x2, 0xfffff000
ADDI x2, x2, 0xfff
```

### PR DESCRIPTION
**Pull request by qwe661234**
After constant propagation and constant folding, we can optimize some ALU or brnach instructions into simple instructions.

Close: #226

### CODE REVIEW

File: src/emulate.c [hunk 1]
```diff
@@ -834,102 +823,102 @@ static bool detect_memcpy(riscv_t *rv, int lib)
     return true;
 }
 
+static bool stdlib_injection(riscv_t *rv, block_t *block)
```
Comment (jserv) [inline src/emulate.c]: Rename to `libc_substitute`.

**Review by jserv — CHANGES_REQUESTED** [review]
Consistently employ the prefix `constopt_` rather than `const_opt` to prevent potential confusion.

File: src/const_opt.c [hunk 2]
```diff
@@ -0,0 +1,1009 @@
+/* RV32I Base Instruction Set */
```
Comment (jserv) [inline src/const_opt.c]: Rename to `src/rv32_constopt.c`.

File: src/emulate.c [hunk 3]
```diff
@@ -957,25 +961,46 @@ static void match_pattern(riscv_t *rv, block_t *block)
...
     }
 }
 
+
+typedef struct const_opt_info {
+    bool is_constant[32];
+    uint32_t const_val[32];
+} const_opt_info_t;
+
+#define CONSTOPT(inst, code)                                              \
```
Comment (jserv) [inline src/emulate.c]: Do `#undef CONSTOPT` when the code generation completes.

**Comment by jserv** [comment]
Both constant folding and constant propagation are related compiler optimization techniques. Constant propagation replaces the bound variable with the constant expression it is bound to. Conversely, constant folding evaluates the expression with all compile-time inputs. I am a bit confused about the proposed changes, as I cannot ensure that a fully functional implementation exists for either constant folding or constant propagation. Can you verify this?

File: src/emulate.c [hunk 4]
```diff
@@ -957,25 +961,46 @@ static void match_pattern(riscv_t *rv, block_t *block)
...
+#include "const_opt.c"
+/* clang-format off */
+static const void *const_opt_table[] = {
+    /* RV32 instructions */
+#define _(inst, can_branch, reg_mask) [rv_insn_##inst] = const_opt_##inst,
+    RV_INSN_LIST
+#undef _
+};
+/* clang-format on */
+typedef void (*opt_func)(rv_insn_t *, const_opt_info_t *);
```
Comment (jserv) [inline src/emulate.c]: Should be `constopt_func_t`.

File: src/emulate.c [hunk 5]
```diff
@@ -957,25 +961,46 @@ static void match_pattern(riscv_t *rv, block_t *block)
...
+static void constant_opt(block_t *block)
```
Comment (jserv) [inline src/emulate.c]: Rename to `optimize_constant` for consistency.

File: src/emulate.c [hunk 6]
```diff
@@ -957,25 +961,46 @@ static void match_pattern(riscv_t *rv, block_t *block)
...
-            break;
             /* TODO: mixture of SW and LW */
             /* TODO: reorder insturction to match pattern */
         }
     }
 }
 
+
+typedef struct const_opt_info {
+    bool is_constant[32];
```
Comment (jserv) [inline src/emulate.c]: Why hard-coded `32`?

File: src/emulate.c [hunk 7]
```diff
@@ -834,102 +823,102 @@ static bool detect_memcpy(riscv_t *rv, int lib)
...
+                            ir->impl = dispatch_table[ir->opcode];
+                            ir->tailcall = true;
+                            return true;
+                        };
+                    }
+                }
+            }
+        }
+        break;
+        /* TODO: inject other frequently-used standard library */
```
Comment (jserv) [inline src/emulate.c]: Rephrase as following:
> Inject other frequently used function calls from the C standard library.

File: src/emulate.c [hunk 8]
```diff
@@ -957,25 +961,48 @@ static void match_pattern(riscv_t *rv, block_t *block)
...
-            }
-            break;
             /* TODO: mixture of SW and LW */
             /* TODO: reorder insturction to match pattern */
         }
     }
 }
 
+
+typedef struct constopt_info {
```
Comment (jserv) [inline src/emulate.c]: Drop the use of `constopt_info`. i.e., use anonymous structure instead.

**Comment by qwe661234** [comment]
> Both constant folding and constant propagation are related compiler optimization techniques. Constant propagation replaces the bound variable with the constant expression it is bound to. Conversely, constant folding evaluates the expression with all compile-time inputs. I am a bit confused about the proposed changes, as I cannot ensure that a fully functional implementation exists for either constant folding or constant propagation. Can you verify this?

There is a simple example in `./build/dhrystone.elf` can be optimized by constant folding and constant propagation.
* Before
```
   0x100f0:	auipc	gp,0x18
   0x100f4:	add	gp,gp,864 
   0x100f8:	add	a0,gp,-1996
   0x100fc:	auipc	a2,0x1a
   0x10100:	add	a2,a2,1384 #
   0x10104:	sub	a2,a2,a0
```
* After
```
   0x100f0:	lui	gp,  0x10108        
   0x100f4:	lui	gp,  0x10468
   0x100f8:	lui	a0, 0xFC9C
   0x100fc:	lui	a2,0x10116
   0x10100:	lui	a2, 0x1067E 
   0x10104:	lui    a2, 0x9e2
```
Moreover, these 6 `lui` instructions can be fused into a fused instruction.

**Comment by jserv** [comment]
> There is a simple example in `build/dhrystone.elf` can be optimized by constant folding and constant propagation.
[...]
> Moreover, these 6 `lui` instructions can be fused into a fused instruction.

Describe this within git commit messages.

File: src/emulate.c [hunk 9]
```diff
@@ -957,25 +962,48 @@ static void match_pattern(riscv_t *rv, block_t *block)
...
-            break;
             /* TODO: mixture of SW and LW */
             /* TODO: reorder insturction to match pattern */
         }
     }
 }
 
+
+typedef struct {
+    bool is_constant[N_RV_REGS];
```
Comment (jserv) [inline src/emulate.c:973]: Does it make sense to check x0, which is always zero?

File: src/emulate.c [hunk 10]
```diff
@@ -957,25 +962,48 @@ static void match_pattern(riscv_t *rv, block_t *block)
...
+#undef _
+};
+#undef CONSTOPT
+
+/* clang-format on */
+typedef void (*constopt_func_t)(rv_insn_t *, constopt_info_t *);
+static void optimize_constant(block_t *block)
+{
+    constopt_info_t constopt_info;
+    memset(&constopt_info, 0, sizeof(constopt_info));
```
Comment (jserv) [inline src/emulate.c]: The C standards guarantee that even if an integer array is located on automatic storage and if there are fewer initializers in a brace-enclosed list then the uninitialized elements must be initialized to 0.

C99 Standard 6.7.8.21
> If there are fewer initializers in a brace-enclosed list than there are elements or members of an aggregate, or fewer characters in a string literal used to initialize an array of known size than there are elements in the array, the remainder of the aggregate shall be initialized implicitly the same as objects that have static storage duration.

You can safely rewrite as following without the invocation of `memset`.
```c
    constopt_info_t info = {0};
```

File: src/emulate.c (hunk 9 above)
Comment (qwe661234) [reply src/emulate.c:973]: Added

**Pull request merged**